### PR TITLE
Replace nl-unit-test with pigweed for credentials

### PR DESCRIPTION
**Pull request by jlatusek**
## Problem
The problem was described in #29342. 

## Changes
These changes are another piece of replacing the nl-unit test with the pigweed-unit test. I changed tests for the credentials. I didn't introduce any changes to the logic of tests; I only replaced one framework with another.

### CODE REVIEW

File: src/credentials/tests/TestGroupDataProvider.cpp [hunk 1]
```diff
@@ -1164,126 +1216,36 @@ void TestGroupDecryption(nlTestSuite * apSuite, void * apContext)
     auto it      = provider->IterateGroupSessions(session_id);
     size_t count = 0, total = 0;
 
-    NL_TEST_ASSERT(apSuite, it);
-    if (it)
+    EXPECT_TRUE(it);
```
Comment (andy31415) [inline src/credentials/tests/TestGroupDataProvider.cpp:1219]: it seems the previous code did the equivalent of ASSERT_TRUE here (i.e. early exit).

```suggestion
    ASSERT_TRUE(it);
```

**Review by andy31415 — APPROVED** [review]

File: src/credentials/tests/TestCommissionerDUTVectors.cpp [hunk 2]
```diff
@@ -49,10 +47,21 @@ static void OnAttestationInformationVerificationCallback(void * context, const D
...
 
-static void TestCommissionerDUTVectors(nlTestSuite * inSuite, void * inContext)
+struct TestCommissionerDUTVectors : public ::testing::Test
+{
+    static void SetUpTestSuite()
+    {
+        CHIP_ERROR error = chip::Platform::MemoryInit();
+        EXPECT_EQ(error, CHIP_NO_ERROR);
+    }
+
```
Comment (arkq) [inline src/credentials/tests/TestCommissionerDUTVectors.cpp:57]:
```suggestion
    static void SetUpTestSuite() { ASSERT_EQ(chip::Platform::MemoryInit(), CHIP_NO_ERROR); }
```

File: src/credentials/tests/TestChipCert.cpp [hunk 3]
```diff
@@ -163,7 +161,17 @@ static void ClearTimeSource(ValidationContext & validContext)
...
 }
 
-static void TestChipCert_ChipToX509(nlTestSuite * inSuite, void * inContext)
+struct TestChipCert : public ::testing::Test
+{
+    static void SetUpTestSuite()
+    {
+        CHIP_ERROR error = chip::Platform::MemoryInit();
+        EXPECT_EQ(error, CHIP_NO_ERROR);
+    }
```
Comment (arkq) [inline src/credentials/tests/TestChipCert.cpp:170]:
```suggestion
    static void SetUpTestSuite() { ASSERT_EQ(chip::Platform::MemoryInit(), CHIP_NO_ERROR); }
```

File: src/credentials/tests/TestDeviceAttestationConstruction.cpp [hunk 4]
```diff
@@ -22,15 +22,25 @@
...
 
-static void TestAttestationElements_Roundtrip(nlTestSuite * inSuite, void * inContext)
+struct TestDeviceAttestationConstruction : public ::testing::Test
+{
+    static void SetUpTestSuite()
+    {
+        CHIP_ERROR error = chip::Platform::MemoryInit();
+        EXPECT_EQ(error, CHIP_NO_ERROR);
+    }
+
```
Comment (arkq) [inline src/credentials/tests/TestDeviceAttestationConstruction.cpp:39]:
```suggestion
    static void SetUpTestSuite() { ASSERT_EQ(chip::Platform::MemoryInit(), CHIP_NO_ERROR); }
```

File: src/credentials/tests/TestDeviceAttestationCredentials.cpp [hunk 5]
```diff
@@ -48,85 +47,95 @@ static const ByteSpan kExpectedPaiPublicKey = DevelopmentCerts::kPaiPublicKey;
...
 } // namespace
 
-static void TestDACProvidersExample_Providers(nlTestSuite * inSuite, void * inContext)
+struct TestDeviceAttestationCredentials : public ::testing::Test
+{
+    static void SetUpTestSuite()
+    {
+        CHIP_ERROR error = chip::Platform::MemoryInit();
+        EXPECT_EQ(error, CHIP_NO_ERROR);
+    }
```
Comment (arkq) [inline src/credentials/tests/TestDeviceAttestationCredentials.cpp:56]:
```suggestion
    static void SetUpTestSuite() { ASSERT_EQ([error](chip::Platform::MemoryInit()), CHIP_NO_ERROR); }
```

File: src/credentials/tests/TestDeviceAttestationCredentials.cpp [hunk 6]
```diff
@@ -48,85 +47,95 @@ static const ByteSpan kExpectedPaiPublicKey = DevelopmentCerts::kPaiPublicKey;
...
+
+TEST_F(TestDeviceAttestationCredentials, TestDACProvidersExample_Providers)
 {
     uint8_t der_cert_buf[kMaxDERCertLength];
     MutableByteSpan der_cert_span(der_cert_buf);
 
     // Make sure default provider exists and is not implemented on at least one method
     DeviceAttestationCredentialsProvider * default_provider = GetDeviceAttestationCredentialsProvider();
-    NL_TEST_ASSERT(inSuite, default_provider != nullptr);
+    EXPECT_NE(default_provider, nullptr);
```
Comment (arkq) [inline src/credentials/tests/TestDeviceAttestationCredentials.cpp:67]:
```suggestion
    ASSERT_NE(default_provider, nullptr);
```

File: src/credentials/tests/TestDeviceAttestationCredentials.cpp [hunk 7]
```diff
@@ -48,85 +47,95 @@ static const ByteSpan kExpectedPaiPublicKey = DevelopmentCerts::kPaiPublicKey;
...
 
     CHIP_ERROR err = default_provider->GetDeviceAttestationCert(der_cert_span);
-    NL_TEST_ASSERT(inSuite, err == CHIP_ERROR_NOT_IMPLEMENTED);
+    EXPECT_EQ(err, CHIP_ERROR_NOT_IMPLEMENTED);
 
     // Replace default provider with example provider
     DeviceAttestationCredentialsProvider * example_dac_provider = Examples::GetExampleDACProvider();
-    NL_TEST_ASSERT(inSuite, example_dac_provider != nullptr);
-    NL_TEST_ASSERT(inSuite, default_provider != example_dac_provider);
+    EXPECT_NE(example_dac_provider, nullptr);
```
Comment (arkq) [inline src/credentials/tests/TestDeviceAttestationCredentials.cpp:74]:
```suggestion
    ASSERT_NE(example_dac_provider, nullptr);
```

File: src/credentials/tests/TestDeviceAttestationCredentials.cpp [hunk 8]
```diff
@@ -48,85 +47,95 @@ static const ByteSpan kExpectedPaiPublicKey = DevelopmentCerts::kPaiPublicKey;
...
-static void TestDACProvidersExample_Signature(nlTestSuite * inSuite, void * inContext)
+TEST_F(TestDeviceAttestationCredentials, TestDACProvidersExample_Signature)
 {
     constexpr uint8_t kExampleMessage[] = { 0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x10, 0x11, 0x12,
                                             0x13, 0x14, 0x15, 0x16, 0x17, 0x20, 0x21, 0x22, 0x23, 0x24, 0x25,
                                             0x26, 0x27, 0x30, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37 };
 
     DeviceAttestationCredentialsProvider * example_dac_provider = Examples::GetExampleDACProvider();
-    NL_TEST_ASSERT(inSuite, example_dac_provider != nullptr);
+    EXPECT_NE(example_dac_provider, nullptr);
```
Comment (arkq) [inline src/credentials/tests/TestDeviceAttestationCredentials.cpp:130]:
```suggestion
    ASSERT_NE(example_dac_provider, nullptr);
```

File: src/credentials/tests/TestDeviceAttestationCredentials.cpp [hunk 9]
```diff
@@ -191,20 +200,20 @@ static void TestDACVerifierExample_AttestationInfoVerification(nlTestSuite * inS
 
     // Make sure default verifier exists and is not implemented on at least one method
     DeviceAttestationVerifier * default_verifier = GetDeviceAttestationVerifier();
-    NL_TEST_ASSERT(inSuite, default_verifier != nullptr);
+    EXPECT_NE(default_verifier, nullptr);
```
Comment (arkq) [inline src/credentials/tests/TestDeviceAttestationCredentials.cpp:203]:
```suggestion
    ASSERT_NE(default_verifier, nullptr);
```

File: src/credentials/tests/TestDeviceAttestationCredentials.cpp [hunk 10]
```diff
@@ -191,20 +200,20 @@ static void TestDACVerifierExample_AttestationInfoVerification(nlTestSuite * inS
...
     AttestationVerificationResult attestationResult = AttestationVerificationResult::kSuccess;
     ByteSpan emptyByteSpan;
     attestationResult = default_verifier->ValidateCertificationDeclarationSignature(ByteSpan(), emptyByteSpan);
-    NL_TEST_ASSERT(inSuite, attestationResult == AttestationVerificationResult::kNotImplemented);
+    EXPECT_EQ(attestationResult, AttestationVerificationResult::kNotImplemented);
 
     DeviceAttestationVerifier * example_dac_verifier = GetDefaultDACVerifier(GetTestAttestationTrustStore());
-    NL_TEST_ASSERT(inSuite, example_dac_verifier != nullptr);
-    NL_TEST_ASSERT(inSuite, default_verifier != example_dac_verifier);
+    EXPECT_NE(example_dac_verifier, nullptr);
```
Comment (arkq) [inline src/credentials/tests/TestDeviceAttestationCredentials.cpp:211]:
```suggestion
    ASSERT_NE(example_dac_verifier, nullptr);
```

File: src/credentials/tests/TestDeviceAttestationCredentials.cpp [hunk 11]
```diff
@@ -258,11 +267,11 @@ static void TestDACVerifierExample_CertDeclarationVerification(nlTestSuite * inS
 
     // Replace default verifier with example verifier
     DeviceAttestationVerifier * example_dac_verifier = GetDefaultDACVerifier(GetTestAttestationTrustStore());
-    NL_TEST_ASSERT(inSuite, example_dac_verifier != nullptr);
+    EXPECT_NE(example_dac_verifier, nullptr);
```
Comment (arkq) [inline src/credentials/tests/TestDeviceAttestationCredentials.cpp:270]:
```suggestion
    ASSERT_NE(example_dac_verifier, nullptr);
```

File: src/credentials/tests/TestDeviceAttestationCredentials.cpp [hunk 12]
```diff
@@ -335,25 +344,25 @@ static void TestDACVerifierExample_NocsrInformationVerification(nlTestSuite * in
                                                  0x8d, 0x85, 0xfb, 0xd7, 0xa0, 0x7c, 0x8e, 0x83, 0x7d, 0xa4, 0xd5, 0xa8, 0xb9 };
 
     DeviceAttestationVerifier * exampleDacVerifier = GetDefaultDACVerifier(GetTestAttestationTrustStore());
-    NL_TEST_ASSERT(inSuite, exampleDacVerifier != nullptr);
+    EXPECT_NE(exampleDacVerifier, nullptr);
```
Comment (arkq) [inline src/credentials/tests/TestDeviceAttestationCredentials.cpp:347]:
```suggestion
    ASSERT_NE(exampleDacVerifier, nullptr);
```

File: src/credentials/tests/TestDeviceAttestationCredentials.cpp [hunk 13]
```diff
@@ -383,7 +392,7 @@ static void TestAttestationTrustStore(nlTestSuite * inSuite, void * inContext)
     };
 
     const AttestationTrustStore * testAttestationTrustStore = GetTestAttestationTrustStore();
-    NL_TEST_ASSERT(inSuite, testAttestationTrustStore != nullptr);
+    EXPECT_NE(testAttestationTrustStore, nullptr);
```
Comment (arkq) [inline src/credentials/tests/TestDeviceAttestationCredentials.cpp:395]:
```suggestion
    ASSERT_NE(testAttestationTrustStore, nullptr);
```

File: src/credentials/tests/TestFabricTable.cpp [hunk 14]
```diff
@@ -163,68 +162,81 @@ static CHIP_ERROR LoadTestFabric_Node02_01(nlTestSuite * inSuite, FabricTable &
...
-void TestLastKnownGoodTimeInit(nlTestSuite * inSuite, void * inContext)
+
+struct TestFabricTable : public ::testing::Test
+{
+
+    static void SetUpTestSuite()
+    {
+        DeviceLayer::SetConfigurationMgr(&DeviceLayer::ConfigurationManagerImpl::GetDefaultInstance());
+        auto error = chip::Platform::MemoryInit();
+        EXPECT_EQ(error, CHIP_NO_ERROR);
```
Comment (arkq) [inline src/credentials/tests/TestFabricTable.cpp:193]:
```suggestion
        ASSERT_EQ(chip::Platform::MemoryInit(), CHIP_NO_ERROR);
```

File: src/credentials/tests/TestFabricTable.cpp [hunk 15]
```diff
@@ -615,56 +627,54 @@ void TestBasicAddNocUpdateNocFlow(nlTestSuite * inSuite, void * inContext)
...
-            NL_TEST_ASSERT(inSuite, fabricTable.PeekFabricIndexForNextAddition(nextFabricIndex) == CHIP_NO_ERROR);
-            NL_TEST_ASSERT_EQUALS(inSuite, nextFabricIndex, 2);
+            EXPECT_EQ(fabricTable.PeekFabricIndexForNextAddition(nextFabricIndex), CHIP_NO_ERROR);
+            EXPECT_EQ(nextFabricIndex, 2);
         }
 
         // Validate contents
         const auto * fabricInfo = fabricTable.FindFabricWithIndex(1);
-        NL_TEST_ASSERT(inSuite, fabricInfo != nullptr);
+        EXPECT_NE(fabricInfo, nullptr);
```
Comment (arkq) [inline src/credentials/tests/TestFabricTable.cpp:654]:
```suggestion
        ASSERT_NE(fabricInfo, nullptr);
```
And you can drop the `if` check in the test.

File: src/credentials/tests/TestFabricTable.cpp [hunk 16]
```diff
@@ -695,74 +705,72 @@ void TestBasicAddNocUpdateNocFlow(nlTestSuite * inSuite, void * inContext)
...
-            NL_TEST_ASSERT(inSuite, fabricTable.PeekFabricIndexForNextAddition(nextFabricIndex) == CHIP_NO_ERROR);
-            NL_TEST_ASSERT_EQUALS(inSuite, nextFabricIndex, 3);
+            EXPECT_EQ(fabricTable.PeekFabricIndexForNextAddition(nextFabricIndex), CHIP_NO_ERROR);
+            EXPECT_EQ(nextFabricIndex, 3);
         }
 
         // Validate contents
         const auto * fabricInfo = fabricTable.FindFabricWithIndex(2);
-        NL_TEST_ASSERT(inSuite, fabricInfo != nullptr);
+        EXPECT_NE(fabricInfo, nullptr);
```
Comment (arkq) [inline src/credentials/tests/TestFabricTable.cpp:757]:
```suggestion
        ASSERT_NE(fabricInfo, nullptr);
```
And drop the `if`. There are more places like that so I will not comment on all of them.

File: src/credentials/tests/TestGroupDataProvider.cpp (hunk 1 above)
Comment (jlatusek) [reply src/credentials/tests/TestGroupDataProvider.cpp:1219]: Fixed

File: src/credentials/tests/TestPersistentStorageOpCertStore.cpp [hunk 17]
```diff
@@ -47,73 +46,83 @@ const ByteSpan kTestIcacSpan{ kTestIcacBuf };
...
 
-void TestAddNocFlow(nlTestSuite * inSuite, void * inContext)
+struct TestPersistentStorageOpCertStore : public ::testing::Test
+{
+    static void SetUpTestSuite()
+    {
+        CHIP_ERROR error = chip::Platform::MemoryInit();
+        EXPECT_EQ(error, CHIP_NO_ERROR);
+    }
+
```
Comment (arkq) [inline src/credentials/tests/TestPersistentStorageOpCertStore.cpp:56]:
```suggestion
    static void SetUpTestSuite() { ASSERT_EQ(chip::Platform::MemoryInit(), CHIP_NO_ERROR); }
```

**Review by arkq — APPROVED** [review]

**Review by bzbarsky-apple — APPROVED** [review]

**Comment by andy31415** [comment]
@jlatusek it seems we have asan complains:

```
==25698==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 424 byte(s) in 1 object(s) allocated from:
    #0 0x55aed38e7e5f in malloc ../../../../../../llvm-llvm-project/compiler-rt/lib/asan/asan_malloc_linux.cpp:68:3
    #1 0x55aed3a57120 in chip::Platform::MemoryAlloc(unsigned long) /__w/connectedhomeip/connectedhomeip/out/sanitizers/../../src/lib/support/CHIPMem-Malloc.cpp:96:12
    #2 0x55aed3a6f14f in chip::Credentials::ChipCertificateSet::Init(unsigned char) /__w/connectedhomeip/connectedhomeip/out/sanitizers/../../src/credentials/CHIPCert.cpp:71:54
    #3 0x55aed39e7a8f in (anonymous namespace)::TestFabricTable_TestUpdateNocFailSafe_Test::PigweedTestBody() /__w/connectedhomeip/connectedhomeip/out/sanitizers/../../src/credentials/tests/TestFabricTable.cpp:1933:9
    #4 0x55aed3920d86 in pw::unit_test::internal::Test::PigweedTestRun() /__w/connectedhomeip/connectedhomeip/out/sanitizers/../../third_party/pigweed/repo/pw_unit_test/light_public_overrides/pw_unit_test/framework_backend.h:701:7
    #5 0x55aed392035a in void pw::unit_test::internal::Framework::CreateAndRunTest<(anonymous namespace)::TestFabricTable_TestUpdateNocFailSafe_Test>(pw::unit_test::internal::TestInfo const&) /__w/connectedhomeip/connectedhomeip/out/sanitizers/../../third_party/pigweed/repo/pw_unit_test/light_public_overrides/pw_unit_test/framework_backend.h:487:20
    #6 0x55aed3a687bf in pw::unit_test::internal::TestInfo::run() const /__w/connectedhomeip/connectedhomeip/out/sanitizers/../../third_party/pigweed/repo/pw_unit_test/light_public_overrides/pw_unit_test/framework_backend.h:654:22
    #7 0x55aed3a681e1 in pw::unit_test::internal::Framework::RunAllTests() /__w/connectedhomeip/connectedhomeip/out/sanitizers/../../third_party/pigweed/repo/pw_unit_test/framework_light.cc:63:13
    #8 0x55aed3a55506 in main /__w/connectedhomeip/connectedhomeip/out/sanitizers/../../third_party/pigweed/repo/pw_unit_test/simple_printing_main.cc:33:10
    #9 0x7fa6fcaf5082 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24082) (BuildId: eebe5d5f4b608b8a53ec446b63981bba373ca0ca)

SUMMARY: AddressSanitizer: 424 byte(s) leaked in 1 allocation(s).
```

**Pull request merged**